### PR TITLE
Import Trezor accounts

### PR DESCRIPTION
**Pull request by spylogsster**
Resolves https://github.com/brave/brave-browser/issues/19000

Import Trezor accounts from device, creating chrome-untrusted://trezor-bridge iframe inside the wallet page/panel and loading the Trezor library inside it. Communication with the library goes via postMessages

Security review updated: https://github.com/brave/security/issues/605#issuecomment-951929000

https://user-images.githubusercontent.com/2965009/138886764-c216709e-0f2d-4013-817a-bd5d591910a3.mp4

## Submitter Checklist:

- [x] I confirm that no security/privacy review [is needed](https://github.com/brave/brave-browser/wiki/Security-reviews), or that I have [requested](https://github.com/brave/security/issues/new/choose) one
- [x] There is a [ticket](https://github.com/brave/brave-browser/issues) for my issue
- [x] Used Github [auto-closing keywords](https://docs.github.com/en/github/managing-your-work-on-github/linking-a-pull-request-to-an-issue) in the PR description above
- [x] Wrote a good [PR/commit description](https://google.github.io/eng-practices/review/developer/cl-descriptions.html)
- [x] Added appropriate labels (`QA/Yes` or `QA/No`; `release-notes/include` or `release-notes/exclude`; `OS/...`) to the associated issue
- [x] Checked the PR locally: `npm run test -- brave_browser_tests`, `npm run test -- brave_unit_tests`, `npm run lint`, `npm run gn_check`, `npm run tslint`
- [x] Ran `git rebase master` (if needed)

## Reviewer Checklist:

- [ ] A security review [is not needed](https://github.com/brave/brave-browser/wiki/Security-reviews), or a link to one is included in the PR description
- [ ] New files have MPL-2.0 license header
- [ ] Adequate test coverage exists to prevent regressions
- [ ] Major classes, functions and non-trivial code blocks are well-commented
- [ ] Changes in component dependencies are properly reflected in `gn`
- [ ] Code follows the [style guide](https://chromium.googlesource.com/chromium/src/+/HEAD/styleguide/c++/c++.md)
- [ ] Test plan is specified in PR before merging

## After-merge Checklist:

- [ ] The associated issue milestone is set to the smallest version that the
  changes has landed on
- [ ] All relevant documentation has been updated, for instance:
  - [ ] https://github.com/brave/brave-browser/wiki/Deviations-from-Chromium-(features-we-disable-or-remove)
  - [ ] https://github.com/brave/brave-browser/wiki/Proxy-redirected-URLs
  - [ ] https://github.com/brave/brave-browser/wiki/Fingerprinting-Protections
  - [ ] https://github.com/brave/brave-browser/wiki/Brave%E2%80%99s-Use-of-Referral-Codes
  - [ ] https://github.com/brave/brave-browser/wiki/Custom-Headers
  - [ ] https://github.com/brave/brave-browser/wiki/Web-Compatibility-Exceptions-in-Brave
  - [ ] https://github.com/brave/brave-browser/wiki/QA-Guide
  - [ ] https://github.com/brave/brave-browser/wiki/P3A

## Test Plan:

- Connect Trezor device
- Go to Accounts import for Wallet
- Try import Trezor accounts

### CODE REVIEW

File: browser/ui/webui/brave_wallet/trezor/trezor_ui.cc
Comment (bridiver) [reply browser/ui/webui/brave_wallet/trezor/trezor_ui.cc:39]: @thypon I know extensions bypass CSP for scripts and I believe they do for CSS as well so I don't think making this more restrictive would prevent anything.

**Review by bridiver — APPROVED** [review]
approving cpp changes, please get approval from @petemill for js/ts

File: components/brave_wallet_ui/common/trezor/trezor_bridge_keyring.ts
Comment (petemill) [inline components/brave_wallet_ui/common/trezor/trezor_bridge_keyring.ts:140]: Seems possible to get overlapping IDs. How about using `window.crypto.randomUUID()`?

File: components/brave_wallet_ui/common/trezor/trezor_bridge_keyring.ts
Comment (petemill) [inline components/brave_wallet_ui/common/trezor/trezor_bridge_keyring.ts:125]: Please use `private` keyword for methods instead of naming convention

File: components/brave_wallet_ui/common/trezor/trezor_bridge_keyring.ts
Comment (petemill) [inline components/brave_wallet_ui/common/trezor/trezor_bridge_keyring.ts:112]: Can we provide a type instead of `any`?

File: components/brave_wallet_ui/common/trezor/trezor_bridge_keyring.ts
Comment (petemill) [inline components/brave_wallet_ui/common/trezor/trezor_bridge_keyring.ts:129]: We don't need `async` keyword if we're explicitly returning a `Promise`

File: components/brave_wallet_ui/common/trezor/trezor_bridge_keyring.ts
Comment (petemill) [inline components/brave_wallet_ui/common/trezor/trezor_bridge_keyring.ts:144]: Seems important here, since we're dealing with crypto, to provide a type for `data` and not `any`

File: components/brave_wallet_ui/common/wallet_api_proxy.js
Comment (petemill) [inline components/brave_wallet_ui/common/wallet_api_proxy.js:85]: This isn't the right file for this `getKeyringsByType` funcion. It's a javascript file (not typescript) meant to interface between our Typescript and the generated Mojom Javascript. We actually also don't need it as now we can directly import mojom struct types and functions from the mojom generated output. But the reason it's not right to expose here is that this isn't even a browser / mojom API.

What's worse is that `getKeyringsByType` has been marked as returning `any` which absolutely defeats the point of type safety for the entire API, and this use case seems like the most important to ensure we minimize bugs by having the compiler find type issues. This was introduced by https://github.com/brave/brave-core/pull/10080.

**I suggest in a follow-up creating a separate module at `common/api/getKeyringsByType.ts` and exporting a function that returns either the browser keyring or one of the hardware keyrings, and this time returns a common interface with `getAccounts`, etc.**

File: components/brave_wallet_ui/trezor/trezor.ts
Comment (petemill) [inline components/brave_wallet_ui/trezor/trezor.ts:8]: doesn't need to be marked `async`

Comment (petemill) [reply components/brave_wallet_ui/trezor/trezor.ts:8]: source should be of type `MessageEventSource`

File: components/brave_wallet_ui/common/trezor/trezor_bridge_keyring.ts
Comment (petemill) [inline components/brave_wallet_ui/common/trezor/trezor_bridge_keyring.ts:140]: please don't use `k[Name]` syntax in ts files for non-global constants.

**Review by petemill — CHANGES_REQUESTED** [review]
I'm glad we found a good architecture for this where we can still use their library.
There seems to be a few issues with type safety in our own code here though. There are very few cases where we would use the `any` type, especially for wallet. On top of some of the other places mentioned in the comments, I think you should specify types for all the messages passing back and forth between the wallet frame and our iframe. I suggest creating a file called `trezor-messages.ts` where you would define and export all the message types (and IDs). Then also in that file you could define functions to send and handle type-safe messages. [Here's an example at the typescript playground](https://www.typescriptlang.org/play?#code/PTAEBcCcFMC8HtIFoC20DO6CGBzDA6cdAKHAE8AHaUAYXhRSwDsATAFUuoF5QByAVyYAbeAGMA1r1AAfPnnBIso0fEFFexUp1r1GrALIZseUDwDexUFdAqGzFgC4dd1hyqXrAS0eh0UT0w4HlaInjgBTn6QAUEAvprkVKAA4tDgAILKqkxEdC4sps56LIaYuNQAZKAW1ja69k688opZaui8ADTBoBRY4AAW6JH+gQDaALrE8VpJAKrCYuJ5xYXL9qXGldXdtsWNgiISGtOJ1GwwCJAAYpBYaGushakZrTnoDwWy84dL9ayaIFAs3Q1AGnnQoAAZoJROBPPAmFDED14H4YhB+tQ0GU8ENiNAAB4URDgKEwuEIlF+Njwc5wRA3O7QAAU4CwkHkTgA6gEWPAAO4dUDYzZOOmXRn3P4sACU21qbI5aXwxL8G3KzJF5RlUwBYHQ9GgSHgA2gyEhniE0EIJBUTD8wqM5Scz0yKjaH0KNWsuwacjSLXdb06PT6gyco14WAACmHeOMhaFwkwnPzeQL8EmAkLvI0AIwAJgAzMdiKrwDTxQzbmhmWnWAKhVq8Dq9aAABL2K2OnEYUBCTziUH9cEOYj1vn8-BYFgsACiADdoDkADLg8DLs3M3jN6Ah5nQBdOdV4RfL8AAHir1xr0A+AD45Vx7-KvJDQAeF-gWH0sPhfY8XBAf6ChKEG6hyt6tRWICTACqA-LUKIzC+FgkLQEIZCgGBRh1PkSDoFQoieBaog9JA8BUJAcIYN0tSHt+v74N08SxK2QA).

image

File: components/brave_wallet_ui/common/wallet_api_proxy.js
Comment (spylogsster) [reply components/brave_wallet_ui/common/wallet_api_proxy.js:85]: https://github.com/brave/brave-browser/issues/19171

File: components/brave_wallet_ui/common/trezor/trezor_bridge_keyring.ts
Comment (spylogsster) [reply components/brave_wallet_ui/common/trezor/trezor_bridge_keyring.ts]: done

File: components/brave_wallet_ui/common/trezor/trezor_bridge_keyring.ts
Comment (spylogsster) [reply components/brave_wallet_ui/common/trezor/trezor_bridge_keyring.ts]: done

File: components/brave_wallet_ui/common/trezor/trezor_bridge_keyring.ts
Comment (spylogsster) [reply components/brave_wallet_ui/common/trezor/trezor_bridge_keyring.ts]: done

File: components/brave_wallet_ui/common/trezor/trezor_bridge_keyring.ts
Comment (spylogsster) [reply components/brave_wallet_ui/common/trezor/trezor_bridge_keyring.ts]: done

File: components/brave_wallet_ui/common/trezor/trezor_bridge_keyring.ts
Comment (spylogsster) [reply components/brave_wallet_ui/common/trezor/trezor_bridge_keyring.ts]: `_createBridge` resolves when iframe is loaded, we need the Promise there, same for `getTrezorAccounts` and `unlock` as they wait for the response form bridge. Removed the Promise for `getAccounts`

File: components/brave_wallet_ui/common/trezor/constants.ts
Comment (spylogsster) [reply components/brave_wallet_ui/common/trezor/constants.ts]: done

File: components/brave_wallet_ui/common/trezor/trezor_bridge_keyring.ts
Comment (petemill) [inline components/brave_wallet_ui/common/trezor/trezor_bridge_keyring.ts:87]: This won't do anything as far as I understand. When you call `Function.bind()` it creates a new function every time. So you're calling to remove event listener that's never been added. However, you don't even need the `.bind(this)` when you add event listener since you're using `=>`  function to preserve `this` scope.

Comment (spylogsster) [reply components/brave_wallet_ui/common/trezor/trezor_bridge_keyring.ts]: done

File: components/brave_wallet_ui/trezor/trezor.ts
Comment (spylogsster) [reply components/brave_wallet_ui/trezor/trezor.ts]: done

File: components/brave_wallet_ui/common/trezor/trezor_bridge_keyring.ts
Comment (spylogsster) [reply components/brave_wallet_ui/common/trezor/trezor_bridge_keyring.ts]: done

File: components/brave_wallet_ui/common/trezor/trezor_bridge_keyring.ts
Comment (spylogsster) [reply components/brave_wallet_ui/common/trezor/trezor_bridge_keyring.ts]: done

File: components/brave_wallet_ui/common/trezor/trezor_bridge_keyring.ts
Comment (spylogsster) [reply components/brave_wallet_ui/common/trezor/trezor_bridge_keyring.ts]: done

File: components/brave_wallet_ui/common/trezor/trezor_bridge_keyring.ts
Comment (darkdh) [inline components/brave_wallet_ui/common/trezor/trezor_bridge_keyring.ts:101]: this fingerprint is here is only first 4 bytes of full identifier and not collision free
```
Note that the fingerprint of the parent only serves as a fast way to detect parent and child nodes in software, and software must be willing to deal with collisions. Internally, the full 160-bit identifier could be used.
```
https://github.com/bitcoin/bips/blob/master/bip-0032.mediawiki

We should use same method of how we get Ledger's device id which is sha256 of first account address

**Review by darkdh — CHANGES_REQUESTED** [review]

File: components/brave_wallet_ui/common/trezor/trezor_bridge_keyring.ts
Comment (spylogsster) [reply components/brave_wallet_ui/common/trezor/trezor_bridge_keyring.ts]: done

**Review by darkdh — APPROVED** [review]
wallet integration lgtm

File: components/brave_wallet_ui/common/async/lib.ts
Comment (petemill) [inline components/brave_wallet_ui/common/async/lib.ts:297]: This function should not be here. This module (and directory) are about dispatching and handling actions with the redux store. Suggest just making a simple module at `common/hardwareDeviceIdFromAddress.ts`

Comment (petemill) [reply components/brave_wallet_ui/common/async/lib.ts:297]: And this doesn't look right:
`await Promise.resolve(crypto.subtle.digest('SHA-256', msgBuffer))` is exactly the same as just writing `crypto.subtle.digest('SHA-256', msgBuffer)`, so this function doesn't need to be `async` at all.

File: components/brave_wallet_ui/trezor/trezor.ts
Comment (petemill) [inline components/brave_wallet_ui/trezor/trezor.ts:19]: Just a question here - why cast between our type and the type from tezor library, why not import `TrezorGetPublicKeyResponse` in to trezor-message.ts just like you have already imported `Unsuccessful` type?

File: components/brave_wallet_ui/common/trezor/trezor-messages.ts
Comment (petemill) [inline components/brave_wallet_ui/common/trezor/trezor-messages.ts:57]: `protected handlers: ...`

File: components/brave_wallet_ui/common/trezor/trezor-messages.ts
Comment (petemill) [inline components/brave_wallet_ui/common/trezor/trezor-messages.ts:71]: `protected removeCommandHandler ...`

File: components/brave_wallet_ui/common/trezor/trezor-messages.ts
Comment (petemill) [inline components/brave_wallet_ui/common/trezor/trezor-messages.ts:82]: `protected abstract onMess....`

File: components/brave_wallet_ui/common/trezor/trezor-messages.ts
Comment (petemill) [inline components/brave_wallet_ui/common/trezor/trezor-messages.ts:94]: "commands"

File: components/brave_wallet_ui/common/trezor/trezor-messages.ts
Comment (petemill) [inline components/brave_wallet_ui/common/trezor/trezor-messages.ts:105]: these can both be `private`

File: components/brave_wallet_ui/common/trezor/trezor-messages.ts
Comment (petemill) [inline components/brave_wallet_ui/common/trezor/trezor-messages.ts:119]: Future improvement ideas (for follow-up and only if you want to do them, don't have to do now):
1. Generate the message ID here, instead of relying on every caller to generate a unique message ID. This function parameter would then be: `command: Omit<TrezorFrameCommand, 'id'>`
2. Make listener optional, or:
3. Instead of accepting a listener callback, add a promise resolve that the caller can await which will even further simplify calling this. Implementation would look roughly like:
```ts
  // T is response type, e.g. UnlockResponse. Resolves as `false` if transport errro
  sendCommandToTrezorFrame = async<T> (command: TrezorFrameCommand): Promise<T | false> => {
    return new Promise<T>(async (resolve) => {
      let bridge = this.getBridge()
      if (!bridge) {
        bridge = await this.createBridge()
      }
      if (!bridge.contentWindow) {
        Promise.resolve(false)
        return
      }
      this.addCommandHandler(command.id, resolve)
      bridge.contentWindow.postMessage(command, this.bridgeFrameUrl)
    })
  }
```

File: components/brave_wallet_ui/common/trezor/trezor-messages.ts
Comment (petemill) [inline components/brave_wallet_ui/common/trezor/trezor-messages.ts:121]: `protected`

File: components/brave_wallet_ui/common/trezor/trezor-messages.ts
Comment (petemill) [inline components/brave_wallet_ui/common/trezor/trezor-messages.ts:161]: `protected`

**Review by petemill — CHANGES_REQUESTED** [review]
Thanks for addressing feedback and separating the hardware keyring logic from the iframe creation and message transport logic. Great job on the tests too.
Just a few final things to query or cleanup. I actually did some of the changes to trezor-messages.ts as I was reviewing as it was complicated for me to follow which methods were private, public or protected. Here's the commit: https://github.com/brave/brave-core/commit/aeec892b8ebbf3d8b6544e6913f6e768fb17c760

On that matter, after adding those keywords and then typescript won't allow access, you may consider just using the exported classes as instances in the trezor keyring (and in the iframe) and calling directly, rather than exporting additional functions to wrap them, as well as the additional wrappers you have in your classes. You could also split the 2 functions in to their own module (e.g. `trezor-bridge-transport.ts` and `trezor-command-handler.ts`, then there's no way to create the wrong class in the wrong environment). Up to you, depends how many messages there will end up being I suppose. If you're going to add more in subsequent PRs, you may want to reduce overhead and go with this way in this PR.

**Review by juliber13 — APPROVED** [review]

File: components/brave_wallet_ui/common/async/lib.ts
Comment (spylogsster) [reply components/brave_wallet_ui/common/async/lib.ts]: moved to new file but keep old version, it throws errors otherwise:
```
      TS2769: No overload matches this call.
  Overload 1 of 5, '(elements: Iterable<number>): Uint8Array', gave the following error.
    Argument of type 'PromiseLike<ArrayBuffer>' is not assignable to parameter of type 'Iterable<number>'.
      Property '[Symbol.iterator]' is missing in type 'PromiseLike<ArrayBuffer>' but required in type 'Iterable<number>'.
  Overload 2 of 5, '(length: number): Uint8Array', gave the following error.
    Argument of type 'PromiseLike<ArrayBuffer>' is not assignable to parameter of type 'number'.
  Overload 3 of 5, '(arrayOrArrayBuffer: ArrayLike<number> | ArrayBuffer | SharedArrayBuffer): Uint8Array', gave the following error.
    Argument of type 'PromiseLike<ArrayBuffer>' is not assignable to parameter of type 'ArrayLike<number> | ArrayBuffer | SharedArrayBuffer'.
      Type 'PromiseLike<ArrayBuffer>' is missing the following properties from type 'SharedArrayBuffer': byteLength, length, slice, [Symbol.species], [Symbol.toStringTag]
```

File: components/brave_wallet_ui/trezor/trezor.ts
Comment (spylogsster) [reply components/brave_wallet_ui/trezor/trezor.ts]: done, thanks!

File: components/brave_wallet_ui/common/trezor/trezor-messages.ts
Comment (spylogsster) [reply components/brave_wallet_ui/common/trezor/trezor-messages.ts]: done

File: components/brave_wallet_ui/common/trezor/trezor-messages.ts
Comment (spylogsster) [reply components/brave_wallet_ui/common/trezor/trezor-messages.ts]: done

File: components/brave_wallet_ui/common/trezor/trezor-messages.ts
Comment (spylogsster) [reply components/brave_wallet_ui/common/trezor/trezor-messages.ts]: done

File: components/brave_wallet_ui/common/trezor/trezor-messages.ts
Comment (spylogsster) [reply components/brave_wallet_ui/common/trezor/trezor-messages.ts]: done

File: components/brave_wallet_ui/common/trezor/trezor-messages.ts
Comment (spylogsster) [reply components/brave_wallet_ui/common/trezor/trezor-messages.ts]: done

File: components/brave_wallet_ui/common/trezor/trezor-messages.ts
Comment (spylogsster) [reply components/brave_wallet_ui/common/trezor/trezor-messages.ts]: done

File: components/brave_wallet_ui/common/trezor/trezor-messages.ts
Comment (spylogsster) [reply components/brave_wallet_ui/common/trezor/trezor-messages.ts]: done

File: components/brave_wallet_ui/common/trezor/trezor-messages.ts
Comment (spylogsster) [reply components/brave_wallet_ui/common/trezor/trezor-messages.ts]: done

File: components/brave_wallet_ui/common/trezor/trezor-bridge-transport.ts
Comment (petemill) [inline components/brave_wallet_ui/common/trezor/trezor-bridge-transport.ts:23]: nit: this doesn't need to be `async` since it doesn't do any `await` (that happens inside another `async` function.

File: components/brave_wallet_ui/common/trezor/trezor-bridge-transport.ts
Comment (petemill) [inline components/brave_wallet_ui/common/trezor/trezor-bridge-transport.ts:34]: I think the new linter coming might complain about this hanging `return` at the end of the function

File: components/brave_wallet_ui/common/trezor/trezor-bridge-transport.ts
Comment (petemill) [inline components/brave_wallet_ui/common/trezor/trezor-bridge-transport.ts:30]:
```
resolve(false)
return
```

**Review by petemill — APPROVED** [review]
Thanks for addressing all the feedback! It's a nice basis to build on. There are still some improvements which could be made in follow-ups as more commands are added, if that happens. 

Please squash before merge, and perhaps address these minor nits if you want. And address this CI issue https://sonarcloud.io/project/security_hotspots?id=brave_brave-core&pullRequest=10681&sinceLeakPeriod=true - is it a false positive?

File: components/brave_wallet_ui/common/trezor/trezor-bridge-transport.ts
Comment (spylogsster) [reply components/brave_wallet_ui/common/trezor/trezor-bridge-transport.ts]: done

File: components/brave_wallet_ui/common/trezor/trezor-bridge-transport.ts
Comment (spylogsster) [reply components/brave_wallet_ui/common/trezor/trezor-bridge-transport.ts]: done

File: components/brave_wallet_ui/common/trezor/trezor-bridge-transport.ts
Comment (spylogsster) [reply components/brave_wallet_ui/common/trezor/trezor-bridge-transport.ts]: done

File: components/brave_wallet_ui/common/trezor/trezor-bridge-transport.ts
Comment (spylogsster) [reply components/brave_wallet_ui/common/trezor/trezor-bridge-transport.ts:30]: keeping current because CI alerts for:
```
resolve(false)
return
```

**Comment by spylogsster** [comment]
the SonarCloud error is false positive and the macos build failed with known issue on mac build stations which is not related to this PR

**Comment by bbondy** [comment]
For historical reference:
Screenshot 2024-02-07 at 9 22 32 AM

**Pull request merged**
